### PR TITLE
Week5 - Redux

### PR DESCRIPTION
**Pull request by ShravanTandel**

### CODE REVIEW

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/vegOnlyCheckbox/VegOnlyCheckbox.js [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
-                    Veg Only
-                </span>
+        <div className="veg-or-nonveg">
+            <div className="checkbox-div">
+                <label>
+                    <input
+                        type="checkbox"
+                        className="checkbox"
+                        value={isVeg}
+                        onClick={() => onCheckboxClick()}
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/vegOnlyCheckbox/VegOnlyCheckbox.js]: you can pass the function onCheckboxClick directly, no need to wrap it inside another function. @ShravanTandel

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/header/Header.js [hunk 2]
```diff
@@ -1,40 +1,55 @@
...
 
-import React from 'react';
+import React, { useContext } from 'react';
+import { ButtonColor } from '../../App';
 
-function Header () {
+function Header({ onClick }) {
+    const color = useContext(ButtonColor);
     return (
         <>
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/header/Header.js]: @ShravanTandel extra fragment tags can be ignored.

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/CartItem.js [hunk 3]
```diff
@@ -1,25 +1,24 @@
...
-                    {cartItem.isVeg ? <div className = 'veg'>Veg</div> : <div className = 'nonVeg'>Non Veg</div>}
-                </div>
-                <div className="dishName">
-                    {cartItem.name}
-                </div>
-                <ActionButtonContainer count={cartItem.count} onClickPlus={onClickPlus} pk={cartItem.pk} onClickMinus={onClickMinus}/>
-                <div className="price">
-                    ₹ {cartItem.price * cartItem.count}
-                </div>
+        <div className="cart_item" pk={cartItem.pk}>
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/CartItem.js]: pk is a very short name, it is not explanatory.

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/components/actionButton/ActionButton.js [hunk 4]
```diff
@@ -1,23 +1,37 @@
...
-        </>
+            <div className="count">{count}</div>
+            <div className="plus" onClick={() => incrementOnPlus(pk)}>
+                +
+            </div>
+        </div>
     );
+}
+
+const mapDispatchToProps = (dispatch) => {
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/components/actionButton/ActionButton.js]: check out bindActionCreators from 'redux'

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js [hunk 5]
```diff
@@ -1,162 +1,181 @@
...
-            })
+import React, { useEffect, useMemo } from 'react';
+import Items from './components/items';
+import { connect } from 'react-redux';
+import updateSearchItemsWithCategoryList from '../../../../actions/searchItemsWithCategoryActionCreator';
+import updateItemsWithCategoryList from '../../../../actions/itemsWithCategoryActionCreator';
+import { menuItemList } from '../../../../models/itemList';
+import { sideMenu } from '../../../../models/menu';
+
+const groupBy = (items) =>
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/groupBy

File: SWIGGYDEMO_REACT/swiggy-react-app/src/App.js [hunk 6]
```diff
@@ -1,39 +1,54 @@
...
+
+    function updateInputText(e) {
+        setSearchInputText(e.target.value);
+    }
+
+    function changeButtonColor(color) {
+        setColor(color === 'green' ? 'red' : 'green');
+    }
+    return (
+        <>
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/App.js]: fragment not needed

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/favouriteCheckbox/FavouriteCheckboxContainer.js [hunk 7]
```diff
@@ -1,10 +1,8 @@
...
-import FavouriteCheckbox from "./FavouriteCheckbox";
+import React from 'react';
+import FavouriteCheckbox from './FavouriteCheckbox';
 
-function FavouriteCheckboxContainer () {
-    return (
-        <FavouriteCheckbox />
-    );
-};
+function FavouriteCheckboxContainer() {
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/favouriteCheckbox/FavouriteCheckboxContainer.js]: no-op component, it is just wrapping another component. Please remove this and use the actual component directly.

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js [hunk 8]
```diff
@@ -1,162 +1,181 @@
...
+    searchItemsWithCategory,
+    isVeg,
+    searchInputText,
+    updateSearchItemsWithCategoryList,
+    updateItemsWithCategoryList,
+}) {
+    const { items, categoryList } = itemsWithCategory;
+    const { searchItems, searchCategoryList } = searchItemsWithCategory;
+
+    let itemsDetails = useMemo(() => {
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: no let, use const instead

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js [hunk 9]
```diff
@@ -1,162 +1,181 @@
...
+        return groupBy(menuItemList);
+    }, [menuItemList]);
+    let categoryListForOriginalItems = sideMenu;
+
+    function updateItemsAndCategoryList() {
+        if (isVeg === true) {
+            const updatedItems = {};
+            const categoryList = [];
+            categoryListForOriginalItems.forEach((cat) => {
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: use const with map like 

const someData = list.map(item => {... compute the new list });

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js [hunk 10]
```diff
@@ -1,162 +1,181 @@
...
-            category.forEach( (i) => {
-                let arrayOfItems = [];
-                arrayOfItems = data.filter( (d) => {
-                    return d.category === i;
-                } )
-                if(arrayOfItems.length != 0) {
-                    arrayOfVegOnlyItems.push(arrayOfItems);
-                    filterCat.push(i);
+        if (searchInputText.length !== 0) {
+            const updatedSearchItems = {};
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: use const and map in same line like 

const data = list.map(....);

you can use map in place of forEach.

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js [hunk 11]
```diff
@@ -1,162 +1,181 @@
...
+            categoryListForOriginalItems.forEach((cat) => {
+                updatedSearchItems[cat] = itemsDetails[cat].filter((item) => {
+                    return (
+                        item.name
+                            .toUpperCase()
+                            .indexOf(searchInputText.toUpperCase()) > -1
+                    );
+                });
+                if (updatedSearchItems[cat].length === 0) {
+                    delete updatedSearchItems[cat];
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: no mutation. create new collection without the needed item.

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js [hunk 12]
```diff
@@ -1,162 +1,181 @@
...
-        onFilter(filterCat);
+        updateItemsWithCategoryList(itemsDetails, categoryListForOriginalItems);
     }
 
-    useEffect( () => {
-        updateItems();
-    }, [isVeg, searchInputText])
+    useEffect(() => {
+        updateItemsAndCategoryList();
+    }, [isVeg, searchInputText]);
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: useMemo can be used to calculate updated item and categories

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js [hunk 13]
```diff
@@ -1,162 +1,181 @@
...
+                            <div className="searchResultEmpty">
+                                Search Result Empty
+                            </div>
+                        )}
+                        <hr />
+                    </>
+                )}
+                {categoryList.map((currentCategory, index) => {
+                    return (
+                        <div id={index + 1} key={index} className="item">
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: are we using the id on the div?

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/components/items/Items.js [hunk 14]
```diff
@@ -1,20 +1,21 @@
...
     return (
         <>
-            {
-                data.map( (d) => {
-                    return (
-                            <Item key={d.pk} item={d} cartItems={cartItems} onClickADD={(onClickADD)} onClickPlus={onClickPlus} onClickMinus={onClickMinus}/>
-                    );
-                })
-            }
+            {data.map((d) => {
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/components/items/Items.js]:
```js
{data.map((d) => (
   <Item key={d.pk} item={d} cartItems={cartItems} />
)}
```

Comment (smtrd3) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/components/items/Items.js]: use one line function expression without using the return keyword when possible

File: SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
+import { menuItemList } from '../models/itemList.js';
+import { produce } from 'immer';
+
+function getItem(items, id) {
+    let item = {};
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js]: no let and for

File: SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        case 'cart/checkedout': {
+            let localCartItems = localStorage.getItem('cart');
+
+            if (localCartItems === null) {
+                localStorage.setItem('cart', JSON.stringify(state));
+            } else {
+                localCartItems = JSON.parse(localCartItems);
+                for (var i = 0; i < state.length; i++) {
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js]: no for and let

File: SWIGGYDEMO_REACT/swiggy-react-app/src/reportWebVitals.js [hunk 17]
```diff
@@ -1,13 +1,15 @@
...
-  if (onPerfEntry && onPerfEntry instanceof Function) {
-    import('web-vitals').then(({ getCLS, getFID, getFCP, getLCP, getTTFB }) => {
-      getCLS(onPerfEntry);
-      getFID(onPerfEntry);
-      getFCP(onPerfEntry);
-      getLCP(onPerfEntry);
-      getTTFB(onPerfEntry);
-    });
-  }
+const reportWebVitals = (onPerfEntry) => {
```
Comment (smtrd3) [inline SWIGGYDEMO_REACT/swiggy-react-app/src/reportWebVitals.js]: remove this we don't need this

**Review by smtrd3 — CHANGES_REQUESTED** [review]
Things to take care

- no unnecessary fragment 
- no for and let 
- code formatting (urgent)

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/vegOnlyCheckbox/VegOnlyCheckbox.js (hunk 1 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/vegOnlyCheckbox/VegOnlyCheckbox.js]: Passed onCheckbox to onClick directly

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/header/Header.js (hunk 2 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/header/Header.js]: Removed

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/CartItem.js (hunk 3 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/CartItem.js]: Changed pk to primaryKey

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/components/actionButton/ActionButton.js (hunk 4 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/cartItems/components/cartItem/components/actionButton/ActionButton.js]: Used bindActionCreator

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js (hunk 5 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: Used _.groupBy method

File: SWIGGYDEMO_REACT/swiggy-react-app/src/App.js (hunk 6 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/App.js]: Removed

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/favouriteCheckbox/FavouriteCheckboxContainer.js (hunk 7 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/Filters/components/favouriteCheckbox/FavouriteCheckboxContainer.js]: Deleted wrapped function

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js (hunk 8 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: Used const

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js (hunk 9 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: Used const with map

Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: Used const with map

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js (hunk 10 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: Used const with map

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js (hunk 11 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: Created new collection

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js (hunk 12 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: Used useMemo

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js (hunk 13 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/ListbyCategory.js]: To scroll to particular div

File: SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/components/items/Items.js (hunk 14 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/components/itemDetailsAndCart/components/listByCategory/components/items/Items.js]: Used one line function expression

File: SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js (hunk 15 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js]: Used const and builtin methods

File: SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js (hunk 16 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/reducers/cartReducer.js]: Used const and forEach

File: SWIGGYDEMO_REACT/swiggy-react-app/src/reportWebVitals.js (hunk 17 above)
Comment (ShravanTandel) [reply SWIGGYDEMO_REACT/swiggy-react-app/src/reportWebVitals.js]: Removed

**Pull request closed without merging**